### ISSUE TITLE
[AMDGPU] LLVM ERROR: Cannot select  (AtomicCmpSwap)

### ISSUE THREAD

**Issue by chichunchen** [llvm/llvm-project#54635]
live demo: https://godbolt.org/z/r1nq3EMx5

Command: llc -mtriple=amdgcn-amd-amdhsa -disable-promote-alloca-to-lds -mcpu=gfx908 -amdgpu-dump-hsa-metadata -filetype=asm  bugpoint-reduced-simplified.ll

Erorr message:
```
LLVM ERROR: Cannot select: t3: i32,ch = AtomicCmpSwap<(load store monotonic monotonic (s32) on `i32 addrspace(5)* undef`, addrspace 5)> t0, undef:i32, undef:i32, undef:i32
  t1: i32 = undef
  t1: i32 = undef
  t1: i32 = undef
```

```
; ModuleID = '<stdin>'
source_filename = "llvm-link"
target datalayout = "e-p:64:64-p1:64:64-p2:32:32-p3:32:32-p4:64:64-p5:32:32-p6:32:32-i64:64-v16:16-v24:32-v32:32-v48:64-v96:128-v192:256-v256:256-v512:512-v1024:1024-v2048:2048-n32:64-S32-A5-G1-ni:7"
target triple = "amdgcn-amd-amdhsa"

define amdgpu_kernel void @"test"() local_unnamed_addr #0 {
entry:
  br label %for.cond

for.cond:                     ; preds = %entry
  br label %while.cond

while.cond:                       ; preds = %for.cond
  br i1 undef, label %for.cond.i74, label %while.body

while.body:                       ; preds = %while.cond
  %0 = cmpxchg i32 addrspace(5)* undef, i32 undef, i32 undef monotonic monotonic, align 4
  unreachable

for.cond.i74:                   ; preds = %for.cond.i74, %while.cond
  br label %for.cond.i74, !llvm.loop !1, !cray.depth !3
}

attributes #0 = { "unsafe-fp-math"="false" }

!llvm.module.flags = !{!0}

!0 = !{i32 7, !"Dwarf Version", i32 4}
!1 = distinct !{!1, !2}
!2 = !{!"llvm.loop.mustprogress"}
!3 = !{i32 1}
```

**Comment by arsenm** [comment]
Private atomics don't make any sense. If we had a target IR verifier, this would fail. Technically we could encode this (at least for the MUBUF path) but there's no reason to really codegen this

**Comment by arsenm** [comment]
I'm guessing you ran into this from the same transform as issue 54274. Either we need to start codegening private atomics using the flat instructions, or we need to stop optimizing flat atomics to private atomics. Probably should go with the codegen route since it makes sense to canonicalize to the true address space

**Comment by arsenm** [comment]
Should be fixed by this stack of patches: https://reviews.llvm.org/D123171

**Comment by arsenm** [comment]
Should be fixed by e6012c8e0fbdf422325e66c96537de03d258deb7
